drafts/zip-whitelisted-transparent-programs.rst

```diff
@@ -0,0 +1,282 @@
+minimal impact on the existing ecosystem.
+
+
+Conventions
+===========
+
+We use the following notation as defined in the Zcash protocol specification
+[#spec-notation]_, reproduced here for convenience:
+
+- length(*S*) means the length of (number of elements in) *S*.
+
+- { *a* .. *b* } means the set or type of integers from *a* through *b* inclusive.
+
+- *a* || *b* means the concatenation of sequences *a* then *b*.
+
+We reproduce the following conversion function from ZIP 32 [#zip-0032]_:
+
+- I2LEOSP\ :sub:`l`\ (*k*) is the byte sequence *S* of length *l*/8 representing in little-endian order the
+  integer *k* in range {0..2\ :sup:`l`\ -1}.
+
+We also define the following helper notation for scripts:
+
+- PUSHDATA(*S*) is a data push of the byte sequence *S* using the standard script data
+  push encoding:
+
+  - I2LEOSP\ :sub:`8`\ (length(*S*)) || *S* if length(*S*) < ``PUSHDATA1``.
+  - ``PUSHDATA1`` || I2LEOSP\ :sub:`8`\ (length(*S*)) || *S* if length(*S*) <= ``0xff``.
+  - ``PUSHDATA2`` || I2LEOSP\ :sub:`16`\ (length(*S*)) || *S* if length(*S*) <= ``0xffff``.
+  - ``PUSHDATA4`` || I2LEOSP\ :sub:`32`\ (length(*S*)) || *S* otherwise.
+
+
+Specification
+=============
+
+Transparent program
+-------------------
+
+A transparent program is defined by four properties:
+
+- A 1-byte program type.
+- A byte sequence ``predicate`` containing a prefix-free encoding describing the
+  predicate.
+- A non-malleable byte sequence ``witness`` containing a prefix-free encoding of evidence
+  satisfying the predicate.
+- A deterministic verification algorithm ``verify_program`` that takes as arguments
+  ``predicate``, ``witness``, and a context object providing deterministic public
+  information about the transaction as well as the block chain (up to and including the
+  block that the transaction is mined in). It returns ``true`` if the predicate is
+  satisfied, and ``false`` otherwise.
+
+ZIPs that define a new transparent program must completely specify all four of these
+properties.
+
+Program types ``{0xfd..0xff}`` are reserved for future extensions to this schema.
+
+The lengths of ``predicate`` and ``witness`` are not required to be constant length, but
+MUST be solely determined by the program type.
+
+Encoding in transactions
+------------------------
+
+A new script opcode ``PROGRAM`` is introduced that redefines the existing ``NOP10``
+opcode.
+
+To create an encumbered transparent output, its ``scriptPubKey`` is set to exactly the
+following:
+
+  PROGRAM PUSHDATA(type || predicate)
+
+When spending an encumbered transparent output, the ``scriptSig`` of the input is set to
+exactly the following:
+
+  PROGRAM PUSHDATA(type || witness)
+
+To a script parser unaware of the ``PROGRAM`` opcode, this results in the following stack,
+which would be interpreted as an "anyone can spend" output:
+
+  PUSHDATA(type || witness) PUSHDATA(type || predicate)
+
+TODO: Alternatively, we could enforce that the script fails-closed for old script
```

Comment:
If breaking old parsers fail is desirable, it would be elegant to repurpose `OP_VERIF` or `OP_VERNOTIF` rather than `OP_NOP10`. These currently cause evaluation to fail if present in a Script (even if in unexecuted branches).